rotkehlchen/externalapis/alchemy.py

```diff
@@ -0,0 +1,424 @@
+                f'Tried to query current price using Alchemy from {from_asset.identifier} '
+                f'to {to_asset.identifier}. But from_asset is not supported in alchemy',
+            )
+            return ZERO_PRICE
+
+        if isinstance(asset_identifier, str):
+            result = self._query(
+                verb='GET',
+                payload={'symbols': asset_identifier},
+                endpoint='by-symbol',
+            )
+        else:  # can only be dict
+            result = self._query(
+                verb='POST',
+                payload={'addresses': [asset_identifier]},
+                endpoint='by-address',
+            )
+        usd_price = self._deserialize_current_price(
+            result=result['data'],
+            asset_identifier=asset_identifier,
+            from_asset=from_asset,
+            to_asset=to_asset,
+        )
+        if usd_price == ZERO_PRICE or to_asset == A_USD:
+            return usd_price
+
+        # The price is in USD, but we need to convert it to the target currency.
+        rate_price = Inquirer.find_price(from_asset=A_USD, to_asset=to_asset)
+        return Price(usd_price * rate_price)
+
+    def query_historical_price(
+            self,
+            from_asset: Asset,
+            to_asset: Asset,
+            timestamp: Timestamp,
+    ) -> Price:
+        """
+        May raise:
+        - NoPriceForGivenTimestamp if no price could be found
+        - PriceQueryUnsupportedAsset if either from_asset or to_asset are not supported
+        """
+        try:
+            from_asset = from_asset.resolve_to_asset_with_oracles()
+            to_asset = to_asset.resolve_to_asset_with_oracles()
+        except UnknownAsset as e:
+            raise PriceQueryUnsupportedAsset(e.identifier) from e
+
+        price_cache_entry = GlobalDBHandler.get_historical_price(
+            from_asset=from_asset,
+            to_asset=to_asset,
+            timestamp=timestamp,
+            max_seconds_distance=DAY_IN_SECONDS,
+            source=HistoricalPriceOracle.ALCHEMY,
+        )
+        if price_cache_entry:
+            return price_cache_entry.price
+
+        try:
+            # if it's a crypto asset use symbol, otherwise use
+            asset_identifier = self._get_asset_identifier(from_asset)
+        except UnsupportedAsset as e:
+            log.warning(
+                f'Tried to query alchemy historical price from {from_asset.identifier} '
+                f'to {to_asset.identifier}. But from_asset is not supported in alchemy',
+            )
+            raise NoPriceForGivenTimestamp(
+                from_asset=from_asset,
+                to_asset=to_asset,
+                time=timestamp,
+            ) from e
+
+        payload: dict[str, str | int] = {
+            'startTime': (start_of_day_timestamp := Timestamp(timestamp - (timestamp % DAY_IN_SECONDS))),  # noqa: E501
+            'endTime': start_of_day_timestamp + DAY_IN_SECONDS,
+        }
```

Comment:
this will query several points right? Can't you set the frequency to 1d? But also, do we want to get the average of prices? Why not set a 1h window around the timestamp that we want? one day seems to much

Reply:
the default frequency is 1d. that's why i didn't include in the payload.

> But also, do we want to get the average of prices? Why not set a 1h window around the timestamp that we want? one day seems to much

yeah, sure.